packages/rxsocketio/src/creators/io.js

```diff
@@ -16,62 +17,86 @@ const defaultOptions = {
+function socketEventReducer(client, obs) {
+  return (acc, [eventName, handler]) => (
+    client.on(eventName, (...args) => obs.next(handler(...args)))
+  );
+}
+
+// emit messages to a topic
+function createTopicEmitter(obs, _actions = actions) {
+  return topic => message => obs.next(_actions.newMessage({topic, message}));
+}
+
+// handle all messages to a topic
+function handleTopic(client, obs, emitToTopic) {
+  return topic => client.on(topic, emitToTopic(topic));
+}
+
+// convert Socket.io client to RxJS Observable
+function createObservableFromSocket(
+  client,
+  topics,
+  _actions = actions,
+  _eventHandlerMap = eventHandlerMap,
+) {
+  const topicBlacklist = Object.keys(_eventHandlerMap);
+  if (topicBlacklist.find(key => topics.indexOf(key) > -1)) {
+    return throwError(new Error(
+      `These topics are reserved and cannot be subscribed to: ${topicBlacklist}`
+    ));
+  }
+  return new Observable(obs => {
+    obs.next(_actions.createClient({client}));
+    // map Socket.io events to Observable events
+    toPairs(_eventHandlerMap).reduce(socketEventReducer(client, obs), undefined);
+    // listen for the topics that the client has subscribed to
+    const emitToTopic = createTopicEmitter(obs);
+    topics.map(handleTopic(client, obs, emitToTopic));
+    client.on('error', error => obs.error(error));
+  });
+}
+
+// Create an observable from a Socket.io client:
 const io = function io({
   url,
   topics = ['message'],
   socketOptions = {},
   stop$ = of(),
   _socketIO = socketIO,
-  _actions = actions
+  _createObservableFromSocket = createObservableFromSocket,
 }) {
-  const ioOptions = {...defaultOptions, socketOptions};
-  const client = _socketIO(url, ioOptions);
-  const action$ = new Observable(obs => {
-    obs.next([client, _actions.createClient({client})]);
-    client.on('connect', () => obs.next(_actions.connect()));
-    client.on('connect_error', error => obs.next(
-      _actions.connectError({error})
-    ));
-    client.on('connect_timeout', timeout => obs.next(
-      _actions.connectTimeout({timeout})
-    ));
-    client.on('disconnect', reason => obs.next(
-      _actions.disconnect({reason})
-    ));
-    client.on('reconnect', attemptNumber => obs.next(
-      _actions.reconnect({attemptNumber})
-    ));
-    client.on('reconnecting', attemptNumber => obs.next(
-      _actions.reconnecting({attemptNumber})
-    ));
-    client.on('reconnect_error', error => obs.next(
-      _actions.reconnectError({error})
-    ));
-    client.on('reconnect_failed', attemptNumber => obs.next(
-      _actions.reconnectFailed({attemptNumber})
-    ));
-    client.on('ping', () => obs.next(
-      _actions.ping(),
-    ));
-    client.on('pong', latency => obs.next(
-      _actions.pong({latency}),
-    ));
-    topics.map(topic => (
-      client.on(topic, message => obs.next(
-        _actions.newMessage({topic, message})
-      ))
-    ));
-    client.on('error', error => obs.error(error));
-  });
```

Comment:
took all this ugly code and moved it to a mapper object instead.  much simpler and more concise.